psl-core/src/main/java/org/linqs/psl/reasoner/sgd/term/SGDObjectiveTerm.java

```diff
@@ -96,40 +95,109 @@ public float evaluate(float[] variableValues) {
         float movement = 0.0f;
-        float weight = rule.getWeight();
+        float variableStep = 0.0f;
+        float newValue = 0.0f;
+        float partial = 0.0f;
 
         GroundAtom[] variableAtoms = termStore.getVariableAtoms();
         float[] variableValues = termStore.getVariableValues();
+        float dot = dot(variableValues);
 
         for (int i = 0 ; i < size; i++) {
             if (variableAtoms[variableIndexes[i]] instanceof ObservedAtom) {
                 continue;
             }
 
-            float dot = dot(variableValues);
-            float gradient = computeGradient(i, dot);
-            float gradientStep = weight * gradient * (learningRate / iteration);
+            partial = computePartial(i, dot, rule.getWeight());
+            variableStep = computeVariableStep(variableIndexes[i], iteration, learningRate, partial,
+                    accumulatedGradientSquares, accumulatedGradientMean, accumulatedGradientVariance,
+                    sgdExtension);
 
-            float newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - gradientStep));
+            newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - variableStep));
             movement += Math.abs(newValue - variableValues[variableIndexes[i]]);
             variableValues[variableIndexes[i]] = newValue;
+
+            if (coordinateStep) {
+                dot = dot(variableValues);
+            }
         }
 
         return movement;
     }
 
-    private float computeGradient(int varId, float dot) {
+    private float computeVariableStep(
+            int variableIndex, int iteration, float learningRate, float partial,
+            Map<Integer, Float> accumulatedGradientSquares,
+            Map<Integer, Float> accumulatedGradientMean,
+            Map<Integer, Float> accumulatedGradientVariance,
+            SGDExtension sgdExtension) {
+        float step = 0.0f;
+        float adaptedLearningRate = 0.0f;
+
+        switch (sgdExtension) {
+            case NONE:
+                step = partial * learningRate;
+                break;
+            case ADAGRAD:
+                if (accumulatedGradientSquares.get(variableIndex) == null) {
+                    accumulatedGradientSquares.put(variableIndex, (float)Math.pow(partial, 2.0f));
+                } else {
+                    accumulatedGradientSquares.put(variableIndex, accumulatedGradientSquares.get(variableIndex)
+                            + (float)Math.pow(partial, 2.0f));
+                }
+
+                adaptedLearningRate = learningRate / (float)Math.sqrt(accumulatedGradientSquares.get(variableIndex) + 1e-8f);
+
+                step = partial * adaptedLearningRate;
+                break;
+            case ADAM:
+                float beta1 = 0.9f * (float)Math.pow(1.0f - 1.0e-8f, iteration - 1.0f);
+                float beta2 = 0.999f;
+                float meanHat = 0.0f;
+                float varianceHat = 0.0f;
```

Comment:
Since you have these variable names which seem to reference some equation(s), can you put the equation in a comment somewhere in this file?
Or, a link to the math.